Use setup.py for Python all installations

bgruening opened the pull request:
Use setup.py for Python all installations instead of the deprecated `easy_install` method.
Using `setup.py` directly is way easier to maintain and not so error prone. `pip` can not (yet) be used as it does not support installation of wheels into user specified directories.

This PR is needed to get the new scientific python stack up and running in Galaxy.

bgruening committed:
Use setup.py for all installations instead of the depreceated easy_install method.

bgruening commented:
Anyone? @natefoo @erasche @davebx

mvdbeek commented:
@bgruening, are you sure pip doesn't support wheels installation into user specified directories?
I was under the impression that --target customdir (https://pip.pypa.io/en/stable/reference/pip_install/?highlight=target#cmdoption--target) package works (this is OSX):

```
(.venv)Marius-MBP:Marius Marius$ pip install -t wheels_in_custom_dir pysam
You are using pip version 7.1.0, however version 7.1.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
Collecting pysam
  Downloading pysam-0.8.3.tar.gz (1.5MB)
    100% |████████████████████████████████| 1.5MB 225kB/s 
Building wheels for collected packages: pysam
  Running setup.py bdist_wheel for pysam
  Stored in directory: /Users/Marius/Library/Caches/pip/wheels/01/54/c3/bfb2e5bb330852e86384fb843c65c61f91197f451ac05c9c2a
Successfully built pysam
Installing collected packages: pysam
Successfully installed pysam-0.8.3
```

If I understand well, there is no binary wheels support for linux yet, but that's not a galaxy issue per se.
Once binary wheels are supported upstream we would automatically benefit without changing galaxy code?!

Also, for installing packages from source a combination of --target and --install_option should work, right?
(https://pip.pypa.io/en/stable/reference/pip_install/?highlight=target#cmdoption--install-option).

Not that I see any disadvantage in using setup.py directly.

bgruening commented:
As far as I know `--target` is not sufficient and my combination of several `--install_option` hacks was not very fruitful. At least in my testing. So the most stable solution I could found that should also be portable is using the setup.py directly. I should have done this from the beginning.

Please let me know if you see any disadvantages in using setup.py. Otherwise I would like to stick to this solution.

mvdbeek commented:
It's true, as far as I can see there are no advantages for using pip over calling setup.py in galaxy's context when installing from source.
The code looks good, I'm testing a few old and new packages and will report back.
One thing, could you edit the docstring of SetupPythonEnvironment on line 1507 to remove/edit the warning about easy_install?

natefoo commented:
FWIW, the support for Linux wheels is currently in progress (by me) but it will be a bit before it's done. The code itself is in pretty good shape, but unfortunately we need to write or modify a PEP to allow it in to pip/wheel.

mvdbeek commented:
I tested package_python_2_7_bcbiogff_0_6_2 (from the main toolshed) with and without the PR, with the PR I get:

```
#############################################
PATH=/Users/Marius/src/test_new_python/tool_dependencies/python/2.7/iuc/package_python_2_7/5caf592bb259/bin:$PATH; export PATH
PYTHONPATH=/Users/Marius/src/test_new_python/tool_dependencies/python/2.7/iuc/package_python_2_7/5caf592bb259:$PYTHONPATH; export PYTHONPATH
PYTHONPATH=/Users/Marius/src/test_new_python/tool_dependencies/python/2.7/iuc/package_python_2_7/5caf592bb259/lib:$PYTHONPATH; export PYTHONPATH
PYTHONPATH=/Users/Marius/src/test_new_python/tool_dependencies/python/2.7/iuc/package_python_2_7/5caf592bb259/lib/python2.7/site-packages:$PYTHONPATH; export PYTHONPATH
PYTHONHOME=/Users/Marius/src/test_new_python/tool_dependencies/python/2.7/iuc/package_python_2_7/5caf592bb259; export PYTHONHOME
PKG_CONFIG_PATH=/Users/Marius/src/test_new_python/tool_dependencies/python/2.7/iuc/package_python_2_7/5caf592bb259/lib/pkgconfig:$PKG_CONFIG_PATH; export PKG_CONFIG_PATH
PATH=$PYTHONHOME/bin:$PATH; export PATH;
                            export PYTHONPATH=/Users/Marius/src/test_new_python/tool_dependencies/bcbiogff/0.6.2/iuc/package_python_2_7_bcbiogff_0_6_2/8adfe232049d/lib/python:$PYTHONPATH;
                            cd /Users/Marius/src/test_new_python/galaxy/database/tmp/tmp-toolshed-mtdj3ilTu/six-1.9.0;
                            python setup.py install --install-lib /Users/Marius/src/test_new_python/tool_dependencies/bcbiogff/0.6.2/iuc/package_python_2_7_bcbiogff_0_6_2/8adfe232049d/lib/python --install-scripts /Users/Marius/src/test_new_python/tool_dependencies/bcbiogff/0.6.2/iuc/package_python_2_7_bcbiogff_0_6_2/8adfe232049d/bin

STDOUT
running install
#############################################

#############################################
PATH=/Users/Marius/src/test_new_python/tool_dependencies/python/2.7/iuc/package_python_2_7/5caf592bb259/bin:$PATH; export PATH
PYTHONPATH=/Users/Marius/src/test_new_python/tool_dependencies/python/2.7/iuc/package_python_2_7/5caf592bb259:$PYTHONPATH; export PYTHONPATH
PYTHONPATH=/Users/Marius/src/test_new_python/tool_dependencies/python/2.7/iuc/package_python_2_7/5caf592bb259/lib:$PYTHONPATH; export PYTHONPATH
PYTHONPATH=/Users/Marius/src/test_new_python/tool_dependencies/python/2.7/iuc/package_python_2_7/5caf592bb259/lib/python2.7/site-packages:$PYTHONPATH; export PYTHONPATH
PYTHONHOME=/Users/Marius/src/test_new_python/tool_dependencies/python/2.7/iuc/package_python_2_7/5caf592bb259; export PYTHONHOME
PKG_CONFIG_PATH=/Users/Marius/src/test_new_python/tool_dependencies/python/2.7/iuc/package_python_2_7/5caf592bb259/lib/pkgconfig:$PKG_CONFIG_PATH; export PKG_CONFIG_PATH
PATH=$PYTHONHOME/bin:$PATH; export PATH;
                            export PYTHONPATH=/Users/Marius/src/test_new_python/tool_dependencies/bcbiogff/0.6.2/iuc/package_python_2_7_bcbiogff_0_6_2/8adfe232049d/lib/python:$PYTHONPATH;
                            cd /Users/Marius/src/test_new_python/galaxy/database/tmp/tmp-toolshed-mtdj3ilTu/six-1.9.0;
                            python setup.py install --install-lib /Users/Marius/src/test_new_python/tool_dependencies/bcbiogff/0.6.2/iuc/package_python_2_7_bcbiogff_0_6_2/8adfe232049d/lib/python --install-scripts /Users/Marius/src/test_new_python/tool_dependencies/bcbiogff/0.6.2/iuc/package_python_2_7_bcbiogff_0_6_2/8adfe232049d/bin

STDERR
error: can't create or remove files in install directory

The following error occurred while trying to add or remove files in the
installation directory:

    [Errno 2] No such file or directory: '/Users/Marius/src/test_new_python/tool_dependencies/bcbiogff/0.6.2/iuc/package_python_2_7_bcbiogff_0_6_2/8adfe232049d/lib/python/test-easy-install-91906.write-test'

The installation directory you specified (via --install-dir, --prefix, or
the distutils default setting) was:

    /Users/Marius/src/test_new_python/tool_dependencies/bcbiogff/0.6.2/iuc/package_python_2_7_bcbiogff_0_6_2/8adfe232049d/lib/python/

This directory does not currently exist.  Please create it and try again, or
choose a different installation directory (using the -d or --install-dir
option).
#############################################
```

Testing this manually, I can see that `pip install -t targetdir .` correctly installs the package,
while a simple `python setup.py --install-dir targetdir` fails.

So this appears to be one advantage of using pip. If you still have a trace of your pip
experiments I could take a look.

EDIT: it does work with python setup.py --install-dir targetdir, locally it was a problem of PYTHONPATH.

mvdbeek commented:
A quick and dirty hack to get package_python_2_7_bcbiogff_0_6_2 back to work is to add `mkdir -p $INSTALL_DIR/lib/python;` on https://github.com/bgruening/galaxy/blob/pip_install/lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py#L1555 before calling setup.py .
Now I don't know if that would be more or less hack-ish than using pip.

bgruening committed:
add mkdir as safeguard and remove old warning

bgruening commented:
@mvdbeek thanks for testing. I added the `mkdir` and removed the warning.
From what I can tell is that `setup.py` is more safe. `pip --install-dir` does not work on every platform in all scenarios. For example see this thread.

http://stackoverflow.com/questions/2915471/install-a-python-package-into-a-different-directory-using-pip

I also remember that the installed lib path can be different if you use `--install-dir` depending on your architecture (32/64 ...) so we would need to take care of this as well.

mvdbeek commented:
@bgruening glad to help. I think your PR works fine now, and if at some point there is a need for pip it can still be implemented. 

I'm :+1: now (I do get a vote nowadays, right?)

On a slightly related note, is it an oversight that setup_python_environment is not listed on https://wiki.galaxyproject.org/ToolDependenciesTagSets ?

bgruening commented:
@mvdbeek we need to add it, but I was still catching up with a new scientific Python stack before I wanted to advertise this feature.
Thanks for reviewing!

jmchilton commented:
@mvdbeek Your detailed testing and comments are super useful. I'd encourage you to vote on things even if we don't officially count them... yet :smile:.
